### DISCUSSION TITLE
Does `assert_type` is supposed to work with `Protocol`?

### DISCUSSION THREAD

**Discussion by rayansostenes** [python/typing#1819] [Q&A] [answered]
Does `assert_type` only supports nominal types?
I'm trying to use it to check a Protocol but it does not work.

```python
from typing import Protocol, assert_type

class MyProtocol(Protocol):
    def my_method(self) -> None:
        ...

class Concrete:
    def my_method(self) -> None:
        ...

x: MyProtocol = Concrete() # OK
assert_type(Concrete(), MyProtocol) # "assert_type" mismatch: expected "MyProtocol" but received "Concrete" 
```

[Pyright Playground Link](https://pyright-play.net/?pyrightVersion=1.1.317&pythonVersion=3.12&strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UAhgM70CmeA%2BvAkwFBekUP0oAWThEwJchQAUYiZQCUALi5RVUACZNgUCHFYQmMABZh1U5hWDyoAWgB8UAHJgUTZWo9QAdD558BUADCLqQghm4qaprauvqGJmYWVrYOzq7unqo%2BXjwAHorCosRklFAAvEEhYTBMUtYAxFAA8gDSXAIsMOyItcEooeF1NCKyJRQNUABEHWwcTJM6SPQQtDCkRgVMuZykNepTI8WSCwBGAK74YaRMSABuTPuTfQM1C0A)

MyPy has a similar error:
```shell
error: Expression is of type "Concrete", not "MyProtocol"  [assert-type]
```

> Also, I tried to find any documentation about how `assert_type` is supposed to work, but I did not found a PEP or something similar explaining it. Does this exists anywhere?

**Comment by JelleZijlstra** [comment, marked as answer]
`assert_type()` works with protocols, but it checks for an exact match, not for assignability. To check that `Concrete` is assignable to MyProtocol, it is enough to do e.g. `x: MyProtocol = Concrete()`.

There is a specification at https://typing.readthedocs.io/en/latest/spec/directives.html#assert-type but it could be made more precise.

**Reply by rayansostenes** [reply]
@JelleZijlstra, thanks for the fast response, I found a nicer alternative to do what I wanted in the first place.
(Give type errors for Protocol implementations at the class definition site, and not only when instantiating the class).

```python
# /tmp/t.py
from typing import Protocol

class MyProtocol(Protocol):
    def my_method(self) -> None:
        raise NotImplementedError


def implements_my_protocol[P: MyProtocol](cls: type[P]) -> type[P]:
    return cls


@implements_my_protocol
class Concrete:
    ...
```

```console
$ npx pyright /tmp/t.py
/tmp/t.py
  /tmp/t.py:13:2 - error: Argument of type "type[Concrete]" cannot be assigned to parameter "cls" of type "type[P@implements_my_protocol]" in function "implements_my_protocol"
    Type "Concrete" is incompatible with type "MyProtocol"
      "Concrete" is incompatible with protocol "MyProtocol"
        "my_method" is not present (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

Although it would nicer if python supported generic upper bounds, Then I could have a generic function that would work for all Protocols, something like:

```python
def implements_protocol[P, T: P](protocol_type: type[P]) -> Callable[[type[T]], type[T]]:
    def wrapper[T: P](cls: type[T]) -> type[T]:
        return cls
    return wrapper
```

In typescript we can do something similar using the `extends` keyword, but there we have
the `implements` keyword that does exactly what I want.

```typescript
interface MyInterface {
    myMethod(): void;
}

class MyClass {}

function implementsInterface<P>() {
    return function <T extends { new (): P }>(target: T) {
        return target;
    };
}
implementsInterface<MyInterface>()(MyClass);
// Argument of type 'typeof MyClass' is not assignable to parameter of type 'new () => MyInterface'.
//   Property 'myMethod' is missing in type 'MyClass' but required in type 'MyInterface'

class MyOtherClass implements MyInterface {}
// Class 'MyOtherClass' incorrectly implements interface 'MyInterface'.
//   Property 'myMethod' is missing in type 'MyOtherClass' but required in type 'MyInterface'
```

**Reply by JelleZijlstra** [reply]
> we have the implements keyword that does exactly what I want

Python has that too! You can add the protocol to your base classes, and the type checker will tell you if you're not implementing the protocol correctly. That does mean you also add a runtime base, though.

**Reply by maflAT** [reply]
> You can add the protocol to your base classes, and the type checker will tell you if you're not implementing the protocol correctly.

This is not entirely true. 
For example, a concrete class with a normal attribute will be compatible with a protocol, that defines the same attribute as a property.
However, if I use that protocol as a base for my concrete class, pyright will complain that I am incorrectly overriding that attribute.

Although I'm not sure if this is intended behavior or an oversight.

**Reply by erictraut** [reply]
> ...pyright will complain that I am incorrectly overriding that attribute. ...I'm not sure if this is intended behavior or an oversight.

This is intended behavior, although it admittedly falls into an area of the typing spec that is currently unspecified. Overload rules are an area I'd like to see us standardize eventually.

I think it would also be a good idea to extend the usage of the `ReadOnly` directive (recently introduced in [PEP 705](https://peps.python.org/pep-0705/)) to protocols so developers don't need to resort to using properties to specify a read-only attribute in protocols. This has always seemed like a hack to me. Properties are descriptor objects. The semantics of a property is similar to that of a simple class variable, but the behaviors diverge when accessed from the class object. For this reason, pyright treats them as incompatible when validating override compatibility.

**Reply by guillaume-mueller** [reply]
> > we have the implements keyword that does exactly what I want
> 
> Python has that too! You can add the protocol to your base classes, and the type checker will tell you if you're not implementing the protocol correctly. That does mean you also add a runtime base, though.

@JelleZijlstra 

What do you mean by "add the protocol to your base classes"? If my concrete class inherits from the my template/protocol class, Pyright doesn't complain if not all my methods are present. I have to do a runtime `x: MyProtocol = Concrete()` to get it complaining.

MRE:

```py
from typing import Protocol

class MyProtocol(Protocol):
    def required_method(self):
        ...

class MyClass(MyProtocol):
    ...

# I have to add this line to get Pyright complaining
_: MyProtocol = MyClass()
```

**Reply by erictraut** [reply]
If you want pyright to report missing methods in `MyClass` in this case, you'd need to mark `MyClass` as `@final`. Without the `@final`, it can be used as an (abstract) base class for other classes, so it's not appropriate to report errors for missing methods unless/until you attempt to instantiate it.

**Reply by guillaume-mueller** [reply]
It works, thanks !
